filter_irrelevant_rules: Do not filter with every literal/regex

IagoAbal opened the pull request:
Due to constant folding, a rule may be able to match a string that does
not actually exist as-is in the target file, but that it is produced by
propagating constants. E.g., a rule may succeed at matching "foobar" if
the target code contains `"foo" + "bar"`, but a "foobar" filter would
skip the file!

Closes PA-3110

test plan:
make test # added new tests
Plus ran p/default-v2 on 46 repos and on average runtime and memory usage were essentially the same. For five repos, there was a more significant slowdown of about 20%, due to more files being analyzed. There is a large increase in "partially scanned files" because more files are being analyzed, and we encounter some parsing errors on them. This is expected and not bad per se.

aryx commented on cli/src/semgrep/constants.py:
```diff
@@ -11,7 +11,7 @@
 DEFAULT_SEMGREP_APP_CONFIG_URL = "api/agent/deployments/scans/config"
 
-DEFAULT_TIMEOUT = 30  # seconds
+DEFAULT_TIMEOUT = 2  # seconds
```
we might want to do that in a separate PR.
It's a pretty important change, deserving its own changelog.

IagoAbal replied:
Agree, see #8879.

aryx commented on src/core/Metavariable.ml:
```diff
@@ -34,6 +34,12 @@ let logger = Logging.get_logger [ __MODULE__ ]
  *)
 type mvar = string [@@deriving show, eq, hash]
 
+module MvarSet = Set.Make (struct
```
You can also use 't Set_.t' (Set_.ml is a polymorphic Set in commons/), or SSet.t (defined in Common.mli I think)

IagoAbal replied:
I am bit fond of defining "new types", so I think it would be nice that `mvar` was not equivalent to `string`, but right now they are plain strings so I'll just go with `StringSet` from `Common2` as suggested.

IagoAbal committed:
filter_irrelevant_rules: Do not filter with every literal/regex

Due to constant folding, a rule may be able to match a string that does
not actually exist as-is in the target file, but that it is produced by
propagating constants. E.g., a rule may succeed at matching "foobar" if
the target code contains `"foo" + "bar"`, but a "foobar" filter would
skip the file!

Closes PA-3110

test plan:
make test # added new tests

IagoAbal committed:
Pad's review

aryx commented on src/optimizing/Analyze_pattern.ml:
```diff
@@ -85,14 +89,35 @@ let extract_strings_and_mvars ?lang any =
-        (* do not recurse there, the type does not have to be in the source *)
-        | TypedMetavar _ -> ()
+        | New (t, _, _, _) when lang = Some Lang.Java ->
+            (* We expect the 'new' keyword to be present.  *)
```
meh, not super useful I think, most Java programs use 'new'.
In the same way if your pattern contains if/then/else/... we usually don't extract the strings for them because
they are too common.

But maybe it helps, I didn't run any benchmark.

IagoAbal replied:
TBH I don't think I benchmarked it, but I'll do now.

aryx commented on src/optimizing/Analyze_pattern.ml:
```diff
@@ -85,14 +89,35 @@ let extract_strings_and_mvars ?lang any =
          * We do now semantic equivance on integers between values so
          * 1000 is now equivalent to 1_000 so we can't "regexpize" it.
          *)
-        | L (String (_, (str, _tok), _)) ->
```
Keep the L (String _) case here, and do nothing, but add a comment saying
we used to extract the string, but we don't because of constant propagation, with the
example you put in the changelog.d

aryx commented on src/optimizing/Analyze_rule.mli:
```diff
@@ -37,7 +37,8 @@ val regexp_prefilter_of_rule :
 
 (* internal, do not use directly, not memoized *)
-val regexp_prefilter_of_formula : Rule.formula -> prefilter option
+val regexp_prefilter_of_formula :
```
maybe explain why the hell we pass an xlang here. It's not super intuitive.

aryx commented on tests/irrelevant_rules/mvar-regex.cs:
```diff
@@ -0,0 +1,9 @@
+{
+    public static void Main()
+    {
+        x.SqlFooBar = string.Format("foobar");
```
maybe a comment here explain why it's good we don't match.

IagoAbal committed:
Pad's review

IagoAbal commented:
Preliminary benchmarks look OK, just ~4% slow in average, a few repos were ~20% slower though. Not many differences in findings, etc. But I think we can fix that by in other ways, just prefer to leave that for subsequent PRs. I'll re-run the benchmarks now after having applied the review comments. If they're good I'll post them and merge.

IagoAbal committed:
Add PA-3110 test
